263-add-pagination-on-social-tab

NLMOTS opened the pull request:
Created pagination in social page of user cards
Logic only in frontend

NLMOTS committed:
Created pagination in social page of user cards

Logic only in frontend

Anatolii-Ivanov commented on frontend/.vscode/settings.json:
```diff
@@ -22,5 +22,15 @@
+        "plusplus",
+        "signalr"
+    ],
+    "[html]": {
+        "editor.defaultFormatter": "vscode.html-language-features"
+    }
```
Let's not touch this file

Anatolii-Ivanov commented on frontend/src/app/modules/social-page/pagination/pagination.component.ts:
```diff
@@ -0,0 +1,39 @@
+    styleUrls: ['./pagination.component.sass'],
+})
+export class PaginationComponent {
+    @Input() lastPage: number = 3;
```
Why 3?

Anatolii-Ivanov replied:
Or what is it?

Anatolii-Ivanov commented on frontend/src/app/modules/social-page/pagination/pagination.component.ts:
```diff
@@ -0,0 +1,39 @@
+        this.onChanged();
+    };
+
+    incCurrent() {
```
check spelling

Anatolii-Ivanov commented on frontend/src/app/modules/social-page/social-page/social-page.component.ts:
```diff
@@ -13,16 +14,61 @@ import { UserFilter } from '../../../models/filters/userFilter';
+
+    currentPage: number = 0;
+
+    paginationPages: number = 10;
```
Why 10? It should go from backend as amount of items...

Anatolii-Ivanov commented on frontend/src/app/modules/social-page/social-page/social-page.component.ts:
```diff
@@ -13,16 +14,61 @@ import { UserFilter } from '../../../models/filters/userFilter';
     onFilterChanges(filters: UserFilter) {
         this.users$ = this.userService.getUsers(filters);
+
+        this.userCount$ = this.users$.pipe(
+            map(users => users.length),
+        );
+
+        this.slicedUsers$ = this.users$.pipe(
+            map((users) => this.chunkArray(users, this.paginationPages)),
+        );
```
Duplication

Anatolii-Ivanov commented on frontend/src/app/modules/social-page/social-page/social-page.component.ts:
```diff
@@ -13,16 +14,61 @@ import { UserFilter } from '../../../models/filters/userFilter';
     }
+
+    onChanged(numberVar: number) {
+        this.currentPage = numberVar - 1;
```
Can't you make it like in one format, so it will be everywhere 1,2,3,4... or 0,1,2,3...

Anatolii-Ivanov commented on frontend/src/app/modules/social-page/social-page/social-page.component.ts:
```diff
@@ -13,16 +14,61 @@ import { UserFilter } from '../../../models/filters/userFilter';
+    }
+
+    private chunkArray(myArray: UserCard[], chunkSize: number): UserCard[][] {
+        const results = [];
+
+        while (myArray.length) {
+            results.push(myArray.splice(0, chunkSize));
+        }
+
+        return results;
```
Do it without mutations

NLMOTS committed:
Resolved conversations

slicing array without mutation
removed duplication
amount of items on page now gets from backend

NLMOTS committed:
Removed comment

Anatolii-Ivanov commented on backend/EasySpeak.Core/EasySpeak.Core.WebAPI/Controllers/UserController.cs:
```diff
@@ -75,5 +75,11 @@ public async Task<ActionResult<List<UserShortInfoDto>>> GetFriends()
+        [HttpGet("amountOfItemsOnPage")]
+        public async Task<ActionResult<int>> GetAmountOfItems()
+        {
+            return Ok(await _userService.GetAmountOfItemsOnPage());
```
Pagination should be done on backend, not getting amount of items

NLMOTS committed:
Implemented pagination on backend

tatianahutii commented on backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:
```diff
@@ -120,6 +121,32 @@ public async Task<List<UserShortInfoDto>> GetFilteredUsers(UserFilterDto userFil
+
+        int usersCount = originaUsers.Count;
+
+        int remainingRange = usersCount - userFilter.PageNumber * AmountOfItemsOnPage;
+
+        int pagesCount = usersCount % AmountOfItemsOnPage == 0 ? usersCount / AmountOfItemsOnPage : usersCount / AmountOfItemsOnPage + 1;
+
+        int startIndex = userFilter.PageNumber * AmountOfItemsOnPage;
+        List<UserShortInfoDto> usersWithPagination;
+
+        if (remainingRange >= (AmountOfItemsOnPage))
+        {
+            usersWithPagination = originaUsers.GetRange(startIndex, AmountOfItemsOnPage);
+        }
+        else
+        {
+            usersWithPagination = originaUsers.GetRange(startIndex, remainingRange);
+        }
```
this should be moved to the DB side

NLMOTS replied:
I didn’t quite understand what needs to be done, there is no complex logic here, there is only one query to the database

tatianahutii replied:
GetFilteredUsers returns a list of all users, and only after they you apply for pagination
But pagination should be applied on the DB side because we don't need to load all users to show only N of them

tatianahutii commented on backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs:
```diff
@@ -336,4 +363,9 @@ public async Task<List<UserShortInfoDto>> GetFriends()
+
+    public async Task<int> GetAmountOfItemsOnPage()
+    {
+        return AmountOfItemsOnPage;
```
fix warning

tatianahutii replied:
everywhere

NLMOTS committed:
Fixed warning

NLMOTS committed:
Completed pagination logic on backend

NLMOTS committed:
Merge branch 'development' into 263-add-pagination-on-social-tab

NLMOTS committed:
Added data folder to git ignore

NLMOTS committed:
Fixed problems with merging
